x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/external/response/ibmwatsonx/IbmWatsonxRankedResponseEntity.java

```diff
@@ -0,0 +1,157 @@
+
+    /**
+     * Parses the Ibm Watsonx ranked response.
+     *
+     * For a request like:
+     *     "model": "rerank-english-v2.0",
+     *     "query": "database",
+     *     "return_documents": true,
+     *     "top_n": 3,
+     *     "input": ["greenland", "google","john", "mysql","potter", "grammar"]
+     * <p>
+     *  The response will look like (without whitespace):
+     *     {
+     *     "rerank": [
+     *       {
+     *         "index": 3,
+     *         "relevance_score": 0.7989932
+     *       },
+     *       {
+     *         "index": 5,
+     *         "relevance_score": 0.61281824
+     *       },
+     *       {
+     *         "index": 1,
+     *         "relevance_score": 0.5762553
+     *       },
+     *       {
+     *         "index": 4,
+     *         "relevance_score": 0.47395563
+     *       },
+     *       {
+     *         "index": 0,
+     *         "relevance_score": 0.4338926
+     *       },
+     *       {
+     *         "index": 2,
+     *         "relevance_score": 0.42638257
+     *       }
+     *   ],
+     *   }
+     *
+     * @param response the http response from ibm watsonx
+     * @return the parsed response
+     * @throws IOException if there is an error parsing the response
+     */
+    public static InferenceServiceResults fromResponse(HttpResult response) throws IOException {
+        var parserConfig = XContentParserConfiguration.EMPTY.withDeprecationHandler(LoggingDeprecationHandler.INSTANCE);
+
+        try (XContentParser jsonParser = XContentFactory.xContent(XContentType.JSON).createParser(parserConfig, response.body())) {
+            moveToFirstToken(jsonParser);
+
+            XContentParser.Token token = jsonParser.currentToken();
+            ensureExpectedToken(XContentParser.Token.START_OBJECT, token, jsonParser);
+
+            positionParserAtTokenAfterField(jsonParser, "results", FAILED_TO_FIND_FIELD_TEMPLATE); // TODO error message
+
+            token = jsonParser.currentToken();
+            if (token == XContentParser.Token.START_ARRAY) {
+                return new RankedDocsResults(parseList(jsonParser, IbmWatsonxRankedResponseEntity::parseRankedDocObject));
+            } else {
+                throwUnknownToken(token, jsonParser);
+            }
+
+            // This should never be reached. The above code should either return successfully or hit the throwUnknownToken
+            // or throw a parsing exception
+            throw new IllegalStateException("Reached an invalid state while parsing the Watsonx response");
+        }
+    }
+
+    private static RankedDocsResults.RankedDoc parseRankedDocObject(XContentParser parser) throws IOException {
+        ensureExpectedToken(XContentParser.Token.START_OBJECT, parser.currentToken(), parser);
+        int index = -1;
+        float score = -1;
+        String documentText = null;
+        parser.nextToken();
+        while (parser.currentToken() != XContentParser.Token.END_OBJECT) {
```

Comment:
Nothing to change for this PR but I think in the future we're trying to use `ObjectParser` or `ConstructingObjectParser` to do this type of parsing instead. Here's an example: https://github.com/elastic/elasticsearch/blob/main/x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/external/response/elastic/ElasticInferenceServiceAuthorizationResponseEntity.java#L50